lisa/tools/tlb_stress.py

```diff
@@ -0,0 +1,1036 @@
+    """Performance metrics for TLB stress testing"""
+
+    # Host metrics
+    dtlb_load_misses_per_sec: float = 0.0
+    itlb_load_misses_per_sec: float = 0.0
+    context_switches_per_sec: float = 0.0
+    tlb_flush_events_per_sec: float = 0.0
+
+    # IPI/Interrupt metrics
+    resched_ipis_delta: int = 0
+    call_function_ipis_delta: int = 0
+
+    # Guest performance metrics
+    throughput_ops_per_sec: float = 0.0
+    p95_latency_ms: float = 0.0
+    p99_latency_ms: float = 0.0
+    steal_time_percent: float = 0.0
+
+    # Timing
+    measurement_duration: float = 0.0
+
+
+@dataclass
+class PerformanceThresholds:
+    """Thresholds for performance degradation detection"""
+
+    # Failure thresholds
+    guest_throughput_degradation_percent: float = 15.0  # >15% drop = fail
+    guest_p95_latency_increase_percent: float = 25.0  # >25% increase = fail
+    guest_steal_time_increase_points: float = 3.0  # >3% absolute increase = fail
+
+    # Warning thresholds
+    host_dtlb_misses_increase_percent: float = 30.0  # >30% increase = warn
+    host_ipis_increase_percent: float = 20.0  # >20% increase = warn
+
+    # Monitoring window
+    perf_measurement_duration: int = 45  # 45 second windows
+
+
+@dataclass
+class SystemMetricsSnapshot:
+    """Lightweight system state snapshot for regression checks"""
+
+    # Performance counters from perf (if available)
+    dtlb_load_misses: float = 0.0
+    itlb_load_misses: float = 0.0
+    context_switches: float = 0.0
+    cpu_migrations: float = 0.0
+
+    # All attributes from PerformanceMetrics to ensure compatibility
+    dtlb_load_misses_per_sec: float = 0.0
+    itlb_load_misses_per_sec: float = 0.0
+    context_switches_per_sec: float = 0.0
+    tlb_flush_events_per_sec: float = 0.0
+
+    # IPI/Interrupt metrics
+    resched_ipis_delta: int = 0
+    call_function_ipis_delta: int = 0
+
+    # Guest performance metrics
+    throughput_ops_per_sec: float = 0.0
+    p95_latency_ms: float = 0.0
+    p99_latency_ms: float = 0.0
+    steal_time_percent: float = 0.0
+
+    # Timing
+    measurement_duration: float = 0.0
+
+    # System state
+    perf_available: bool = False
+    interrupts_snapshot: str = ""
+    telemetry_path: str = ""
+
+
+class TlbStress(Tool):
+    """
+    TLB stress testing tool with performance monitoring capabilities.
+
+    Provides core functionality for TLB stress testing including:
+    - Custom C program deployment and execution
+    - Performance metric collection (dTLB misses, steal time, etc.)
+    - Environment detection (guest vs host)
+    - Threshold-based degradation analysis
+    """
+
+    # Detection - absolute paths and consistent naming
+    _bin = "/usr/local/bin/tlbstress"
+    _work = "/opt/tlb_flush"
+
+    def __init__(self, node: "Node", *args: Any, **kwargs: Any) -> None:
+        super().__init__(node, *args, **kwargs)
+        self._is_guest_environment: Optional[bool] = None
+        self._performance_thresholds = PerformanceThresholds()
```

Comment:
This variable is used only in one method, it doesn't need to be an instance level variable.

Reply:
done

Reply:
its now updated properly.